armi/reactor/blocks.py

```diff
@@ -1580,61 +1580,80 @@ def setPinPowers(
-            The "ARMI pin ordering" is used, which is counter-clockwise from 3 o'clock.
+        jmax: list of ints, required
+            Number of pins per ring. Element 0 is the number of pins in ring 1, element 1
+            is the number of pins in ring 2, and so on. This parameter should probably be
+            removed because it can be derived from the block.
+
+        removeSixCornerPins: bool, optional.
+            Defaults to False
+
+        powerKeySuffix: str, optional
+            Must be either an empty string, :py:const:`NEUTRON <armi.physics.neutronics.const.NEUTRON>`,
+            or :py:const:`GAMMA <armi.physics.neutronics.const.GAMMA>`. Defaults to empty
+            string.
+
+        Notes
+        -----
+        This method can handle assembly rotations by using the *pinLocation* parameter.
         """
-        self.p.pinPowers = numpy.zeros(numPins)
-        self.p["linPowByPin" + powerKeySuffix] = numpy.zeros(numPins)
+        powerKey = f"linPowByPin{powerKeySuffix}"
+        self.p[powerKey] = numpy.zeros(numPins)
+
+        # Loop through rings
         j0 = jmax[imax - 1] / 6
         pinNum = 0
-        for i in range(imax):  # loop through rings
-            for j in range(jmax[i]):  # loop through positions in ring i
+        for i in range(imax):
+            # Loop through positions in ring i
+            for j in range(jmax[i]):
                 if removeSixCornerPins and i == imax - 1 and math.fmod(j, j0) == 0.0:
                     linPow = 0.0
                 else:
+                    # TODO: This appears to need fixing to account for blocks in fueled
+                    # TODO: assemblies that contain elevations with pins but no fuel,
+                    # TODO: such as for an axial shield. These blocks should still have
+                    # TODO: powers that are eligible for rotation.
                     if self.hasFlags(Flags.FUEL):
-                        # -1 to map from pinLocations to list index
+                        # -1 is needed in order to map from pinLocations to list index
                         pinLoc = self.p.pinLocation[pinNum] - 1
                     else:
                         pinLoc = pinNum
                     linPow = powers[pinLoc]
-                self.p.pinPowers[pinNum] = linPow
-                self.p["linPowByPin" + powerKeySuffix][pinNum] = linPow
+                self.p[powerKey][pinNum] = linPow
                 pinNum += 1
 
-        if powerKeySuffix == GAMMA:
-            self.p.pinPowersGamma = self.p.pinPowers
-        elif powerKeySuffix == NEUTRON:
-            self.p.pinPowersNeutron = self.p.pinPowers
-
-        if gamma:
-            self.p.pinPowers = self.p.pinPowersNeutron + self.p.pinPowersGamma
-        else:
-            self.p.pinPowers = self.p.pinPowersNeutron
+        # If using the *powerKeySuffix* parameter, we also need to set total power. If
+        # you were setting gamma power, the logic below assumes that the neutron powers
```

Comment:
This logic assumes that `setPinPowers()` will always be called in order: first with `powerKeySuffix == NEUTRON`, then with `powerKeySuffix == GAMMA`. I think it would be more robust to implement in a way that will give you the correct total power regardless of the order in which `setPinPowers()` is called. Maybe we could just check for the existence of each type of power and add it to the total if it exists?

Reply:
Agreed; the existing code assumes that you'll do the operations sequentially. I guess the real problem is with with gamma. The code will abort at least, since it's trying to add None + np.ndarray().

if powersuffix == "":
   # assume pure neutron
   total = neutron values
   neutron = neutron values
   gamma = None (should we set to 0.0?...probably not, since we'd may break others depending on seeing None.
elif powersuffix == NEUTRON:
   # assume pure neutron:
   same as powersuffix = ""
else:
   # coupled
   gamma = gamma values
   assert neutron values are present
   total = neutron + gamma

Reply:
I think that makes sense

Reply:
Both the module and test has been updated to address Mike's suggestion.